### PR TITLE
[#1221] improvement(common): Add the method `checkValue` for the ConfigEntry

### PR DESCRIPTION
**Pull request by qqqttt123**
### What changes were proposed in this pull request?
Now, ConfigEntry lacks the ability of check config options. So we add the support of this feature referring to Spark.

### Why are the changes needed?

Fix: #1221 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Add new ut.

### CODE REVIEW

**Comment by jerryshao** [comment]
Probably be better to add the `checkValue` to all the configurations that can be checked, not only the string configurations.

**Comment by jerqi** [comment]
> Probably be better to add the `checkValue` to all the configurations that can be checked, not only the string configurations.

Added.

**Comment by qqqttt123** [comment]
> > Probably be better to add the `checkValue` to all the configurations that can be checked, not only the string configurations.
> 
> OK, I will try to add more checks.

Added.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.config;
+
+public interface ConfigConstants {
+  String NOT_BLANK_ERROR_MSG = "The value can't be the blank";
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java:9]: "The value can't be blank"

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java [hunk 2]
```diff
@@ -136,6 +163,13 @@ public ConfigEntry<Optional<T>> createWithOptional() {
     // Unless explicitly set by the user, null values are not expected to occur.
     conf.setStringConverter(t -> t.map(stringConverter).orElse(null));
     conf.setOptional();
+    conf.setValidator(
+        optionValue -> {
+          if (validator == null) {
+            return;
+          }
+          optionValue.ifPresent(value -> validator.accept(value));
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:171]: Can you please optimize the logic here? It is a bit confusing to have both null and optional check here.

Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:171]: if optionValue is empty, we will skip the check.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java (hunk 1 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java:9]: OK.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java [hunk 3]
```diff
@@ -43,12 +42,10 @@ public class IcebergTableOps implements AutoCloseable {
...
     String catalogType = icebergConfig.get(IcebergConfig.CATALOG_BACKEND);
     if (!IcebergCatalogBackend.MEMORY.name().equalsIgnoreCase(catalogType)) {
-      Preconditions.checkArgument(
-          StringUtils.isNotBlank(icebergConfig.get(IcebergConfig.CATALOG_WAREHOUSE)),
-          "Catalog warehouse can't be blank");
-      Preconditions.checkArgument(
-          StringUtils.isNotBlank(icebergConfig.get(IcebergConfig.CATALOG_URI)),
-          "Catalog uri can't be blank");
+      // The method `get` will call the method `checkValue`, just check the values of the config
+      // options
```
[the remark this answers is not in the document]
Comment (qqqttt123) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java:46]: OK. I will remove `The method `get` will call the method `checkValue`.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java (hunk 2 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:171]: I will add some comments here.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (qqqttt123) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java:46]: I got your point. I will change the comment
```
The method `get` will check the values of config options.
```

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java (hunk 2 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:172]: Fixed.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java (hunk 1 above)
Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java]: Fixed.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (qqqttt123) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java]: Fixed.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.config;
+
+public interface ConfigConstants {
+  String NOT_BLANK_ERROR_MSG = "The value can't be blank";
```
Comment (diqiu50) [inline common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java:9]: I think using 'Missing config' can more clearly indicate the error.

Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigConstants.java:9]: If you configure " ",  the validator will report error, too.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java [hunk 5]
```diff
@@ -42,8 +42,7 @@ public class IcebergTableOps implements AutoCloseable {
   public IcebergTableOps(IcebergConfig icebergConfig) {
     String catalogType = icebergConfig.get(IcebergConfig.CATALOG_BACKEND);
     if (!IcebergCatalogBackend.MEMORY.name().equalsIgnoreCase(catalogType)) {
-      // The method `get` will call the method `checkValue`, just check the values of the config
-      // options
+      // The method `get` will check the values of config options.
```
Comment (jerryshao) [inline catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java:45]: You don't have to add the comments here, it's not necessary.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/utils/IcebergCatalogUtil.java [hunk 6]
```diff
@@ -45,8 +45,7 @@ private static JdbcCatalog loadJdbcCatalog(Map<String, String> properties) {
     IcebergConfig icebergConfig = new IcebergConfig(properties);
     String driverClassName = icebergConfig.getJdbcDriver();
 
-    // The method `get` will call the method `checkValue`, just check the values of the config
-    // options
+    // The method `get` will check the values of config options.
```
Comment (jerryshao) [inline catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/utils/IcebergCatalogUtil.java:48]: Neither here.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java (hunk 5 above)
Comment (qqqttt123) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/ops/IcebergTableOps.java]: Fixed.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/utils/IcebergCatalogUtil.java (hunk 6 above)
Comment (qqqttt123) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/utils/IcebergCatalogUtil.java]: Fixed.

File: common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java (hunk 2 above)
Comment (jerryshao) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java:172]: I think you can change like this:

```
Stream.of(Optional.ofNullable(validator), optionValue).allMatch(Optional::isPresent) {
validator.accept(value)
```

Comment (qqqttt123) [reply common/src/main/java/com/datastrato/gravitino/config/ConfigEntry.java]: Fixed.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**